server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -430,7 +430,15 @@ private static void parseObject(final DocumentParserContext context, String curr
             context.path().setWithinLeafObject(false);
             context.path().remove();
         } else {
-            parseObjectDynamic(context, currentFieldName);
+            DynamicTemplate dynamicTemplate = context.findDynamicTemplate(currentFieldName, DynamicTemplate.XContentFieldType.OBJECT);
+            if (context.parent().subobjects() || dynamicTemplate != null) {
+                parseObjectDynamic(context, currentFieldName);
+            } else {
+                context.parser().nextToken(); // Skipping Object-start
+                context.path().addDottedFieldName(currentFieldName);
+                innerParseObject(context);
```

Comment:
Thanks for pointing this out, my reasoning here is that at this stage in parsing we are not sure if the object we are parsing contains properties, other nested objects or both therefore we rely on the innerParseObject switch-case to kind-of-recursively unpack objects and parse values. This, together with the fact that we add the dottedFileName before the call and remove it right after, works seamlessly with this logic. 
Basically I am using dottedFieldName as a stack by pushing to it field names while we encounter nested objects and popping from it when we finish parsing any of the nested objects.

After further inspection I think that we can call `parseObjectOrNested` instead of `innerParseObject` after adjusting the context by adding the dottedFieldName. This will maintain the same logic I implemented (since inside `parseObjectOrNested` we call `innerParseObject`) but the `parseObjectOrNested` method will handle all the checks and skip the object start that I was doing manually here.